integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java

```diff
@@ -776,33 +793,127 @@ private int chownInternal(String path, long uid, long gid) {
-   * only sequentially, and never be modified in Alluxio, we delete the existing file
-   * and create a new file for writing in open() and consider truncate() to size 0 as
-   * a noop to fulfill the overwrite requirement.
+   * Since files can be written only once, only sequentially,
+   * and never be modified in Alluxio, truncate is not supported internally by Alluxio.
+   *
+   * In Alluxio Fuse, we support truncate in some special cases.
    *
    * @param path the file to truncate
    * @param size the size to truncate to
    * @return 0 if succeed, error code otherwise
    */
   @Override
   public int truncate(String path, long size) {
-    LOG.debug("truncate {} to {}", path, size);
+    return AlluxioFuseUtils.call(LOG, () -> truncateInternal(path, size),
+        "Fuse.Truncate", "path=%s,size=%d", path, size);
+  }
+
+  private int truncateInternal(String path, long size) {
+    // Truncate scenarios:
+    // 1. Truncate size = 0, file does not exist => no-op
+    // 2. Truncate size = 0, file exists and completed
+    // => noop if file size = 0 or delete file
+    // TODO(lu) delete open file entry if any
+    // 3. Truncate size = 0, file exists and is being written by current Fuse
+    // => noop if written size = 0, otherwise delete file and update create file entry
+    // 4. Truncate size = 0, file exists and is being written by other applications
+    // => error out
+    // 5. Truncate size != 0, file does not exist => error out
+    // 6. Truncate size != 0, file is completed
+    // => no-op if file size = truncate size, error out otherwise
+    // 7. Truncate size != 0, file is being written by other applications
+    // => error out, don't know exact file written size
+    // 8. Truncate size != 0, file is being written by current Fuse
+    // => no-op if written size = truncate size, error out otherwise
+    // Error out in all other cases
+    final AlluxioURI uri = mPathResolverCache.getUnchecked(path);
+    URIStatus status = null;
+    try {
+      status = mFileSystem.getStatus(uri);
+    } catch (FileDoesNotExistException | InvalidPathException e) {
+      status = null;
+    } catch (Throwable t) {
+      LOG.error("Failed to truncate path {} to {}. Failed to get file status", path, size, t);
+      return -ErrorCodes.EIO();
+    }
+    CreateFileEntry<FileOutStream> ce = mCreateFileEntries.getFirstByField(PATH_INDEX, path);
     if (size == 0) {
-      // truncate may be called in overwrite process:
-      // open(openflag=0b2) - truncate to size 0 - write - flush - release
-      if (!mCreateFileEntries.contains(PATH_INDEX, path)) {
-        LOG.error("Cannot truncate {} to {}. The file is not opened for overwrite", path, size);
+      if (status == null) {
+        // Case 1: Truncate size = 0, file does not exist => no-op
+        return 0;
+      }
+      try {
+        // Case 2: Truncate size = 0, file exists and completed
+        if (status.isCompleted()) {
```

Comment:
This looks good

Reply:
The following write will error out because file is already deleted